fvm/transaction.go

```diff
@@ -77,14 +80,66 @@ func (i *TransactionInvocator) Process(
 
 	location := common.TransactionLocation(proc.ID[:])
 
+	var predeclaredValues []runtime.ValueDeclaration
+
+	if ctx.AccountFreezeAvailable {
+
+		setAccountFrozen := runtime.ValueDeclaration{
```

Comment:
You probably want to declare and `Available` function here, as otherwise this function is available in all Cadence programs – which means **anyone** could freeze **any** account!

This `setAccountFrozen` function should probably only be available in the service account contract, right? In that case, the availability function should return true if the location parameter is an address location, where the address is the service account address.

Reply:
You are right, however I don't think this logic (who can freeze) belong to this type. This type enables the function based on boolean. The logic of who sets this boolean (ie. account is service one) is simply external.

Reply:
I assumed that we would want to have a function in the service account that would allow calling this function (and probably emit and event). It seems very dangerous to expose this to all Cadence code when it is enabled. 

I highly recommend making this function only available to certain Cadence code. I agree that it shouldn't be hardcoded to the service account and it might be best if it is configurable, so maybe change the FVM option from a boolean to e.g. an optional address.

Reply:
That's exactly what this does. But instead of having two pieces of logic in one place (deciding who can freeze and enabling it for Cadence) this functions only does the latter, while decision (ie. only for service account or so) it made by caller (or whoever provides the context)

Reply:
How can the function that determines who can call the freezing function be implemented? I think I'm worried because I can't think of a way it can be implemented

Reply:
Well, its my fault. That was part of the plan, but I actually forgot to implement it 🦆 
Idea is simple - we add another transaction processor in the chain, which checks this and sets the flag on context.
Sorry about that

Reply:
No worries! I fear this will be a security problem though: When a transaction processor is running a transaction where this freezing function is enabled, than it will be available to all code. If (hypothetically) that transaction imports some code that it trusts, and it gets compromised, than it might call the freezing function. 

I highly recommend limiting the availability of the function when it is enabled instead of injecting it to all Cadence programs, because I think that's safer.

Reply:
@turbolent I just added this missing functionality. Hope it all makes sense now!